Update README.md and create FAQ.md (#83)
openpubkey/openpubkey: README.md

# OpenPubkey

## Overview

OpenPubkey is a protocol for leveraging OpenID Providers (OPs) to bind identities to public keys. It adds user- or workload-generated public keys to [OpenID Connect (OIDC)](https://openid.net/developers/how-connect-works/), enabling identities to sign messages or artifacts under their OIDC identity. 

We represent this binding as a PK Token. This token proves control of the OIDC identity and the associated private key at a specific time, as long as a verifier trusts the OP. Put another way, the PK Token provides the same assurances as a certificate issued by a Certificate Authority (CA) but critically, does not require adding a CA. Instead, the OP fulfills the role of the CA. This token can be distributed alongside signatures in the same way as a certificate.

OpenPubkey does not add any new trusted parties beyond what is required for OpenID Connect. It is fully compatible with existing OpenID Providers (Google, Azure/Microsoft, Okta, OneLogin, Keycloak) without any changes to the OpenID Provider.

> OpenPubkey can be used to enable a myriad of applications. For example, companies building on OpenPubkey include:

Reviewer: Can we cut this sentence? It's correct, flows and reasonable but I want the readme overview to be as concise as possible and it doesn't add anything.

Not a blocker, this PR can be merged without cutting this.

Reviewer: ```suggestion
For example, companies building on OpenPubkey include:
```

Author: Fixed

* [Docker, Inc](https://www.docker.com/) is building a public container registry [with OpenPubkey used to sign Docker Official Images](https://www.docker.com/blog/signing-docker-official-images-using-openpubkey/).

* [BastionZero](https://www.bastionzero.com/) uses OpenPubkey to provide secure remote access to infrastructure.

> OpenPubkey is a Linux Foundation project open-sourced under the Apache 2.0 license. This project presently provides an OpenPubkey client and verifier for creating and verifying PK Tokens from Google’s OP (for users) and GitHub’s OP (for workloads).

Reviewer: ```suggestion
OpenPubkey is a Linux Foundation project. It is open source and licensed under the Apache 2.0 license. This project presently provides an OpenPubkey client and verifier for creating and verifying PK Tokens from [Google’s OP (for users)](https://developers.google.com/identity/openid-connect/openid-connect) and [GitHub’s OP (for workloads)](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect).
```

Author: Fixed

## How Does OpenPubkey Work?

### OpenPubkey and User Identities

> OpenPubkey supports both workload identities and user identities. Let's look at how this works for users and then show how to extend OpenPubkey to workloads.

Reviewer: If we are going to include the subheading `### OpenPubkey and User Identities` this I propose we cut my two sentences here. I only kept them because I cut the subheading.

```suggestion

```

Author: Fixed. Meant for this sentence to be directly under the major header. Thanks for the catch, Ethan

In OpenID Connect (OIDC) users authenticate to an OP (OpenID Provider), and the OP grants the user an ID Token. These ID Tokens are signed by the OP and contain claims made by the OP about the user such as the user's email address. Important to OpenPubkey is the `nonce` claim in the ID Token.

The `nonce` claim in the ID Token is a random value sent to the OP by the user's client during authentication with the OP. OpenPubkey follows the OpenID Connect authentication protocol with the OP, but it transmits a `nonce` value set to the cryptographic hash of both the user's public key and a random value so that the `nonce` is still cryptographically random, but any party that speaks OpenPubkey can check that ID Token contains the user's public key. From the perspective of the OP, the `nonce` looks just like a random value.

> Let's look at an example where a user, Alice, leverages OpenPubkey to get her OpenID Provider, `google.com`, to bind her OIDC identity, `alice@acme.co`, to her public key `alice-pubkey`. To do this, Alice starts her OpenPubkey client.

Author: Fixed!

# OpenPubkey

## Overview

> OpenPubkey allows any OpenID Connect (OIDC)-speaking identity provider to attest to cryptographic public keys alongside identities. This establishes cryptographic signing keys at scale for users and workloads, in a manner that is scalable for workloads and seamless and invisible to users. OpenPubkey does not require any changes to OpenID Connect so it works with today’s OpenID providers.

Reviewer: Thanks for making this PR. The first few sentence of the readme are the most important as it tells tell someone visiting this repo what OpenPubkey does and why they should care. We should make every word count. I think we should  workshop this a bit.

While they are far from perfect I do prefer the original two sentences as I think they do a better job communicating what OpenPubkey does.

"OpenPubkey adds user or workload generated public keys to OpenID Connect (OIDC) enabling identities to sign messages or artifacts under their OIDC identity. In essence, OpenPubkey is a protocol for getting OpenID Providers (OPs) to bind identities to public keys."

The new description doesn't seem to be as clear. For instance:

> OpenPubkey allows any OpenID Connect (OIDC)-speaking identity provider to attest to cryptographic public keys alongside identities.

* Whose public keys? A user? An OP? 

* Why the word cryptographic here? 

* The OP doesn't know about the public key so it can't attest to the public key.

* It reads me like this is a library for Identity Providers, not parties relying on identity providers.

I'm not sure what the following sentence adds. I would cut it.

> This establishes cryptographic signing keys at scale for users and workloads, in a manner that is scalable for workloads and seamless and invisible to users.

Author: Thanks for the collaboration on this Ethan. Updated the "overview" and "how does this work" sections. [Update README.md](https://github.com/openpubkey/openpubkey/pull/83/commits/85ca102ebe604104a3bfdaeeb88dcf6e05c794df)

OpenPubkey works by essentially turning an OpenID Provider (OP) into a Certificate Authority (CA). (A CA is a trusted entity that issues certificates that cryptographically bind an identity with a cryptographic public key.) To do this, OpenPubKey introduces a new cryptographic object — the PK Token — that binds an identity (i.e., a user or a workload) to its public key.

The PK Token and the public key can then be used to enable a myriad of applications. For example:

> ✍️ [Docker, Inc](https://www.linkedin.com/company/docker/) will use OpenPubKey to sign containers in a public container registry.

Reviewer 2: What is the intent of this character  ✍️?

Why link to company LinkedIn pages rather than the company home pages?

Author: It was meant to reference the "sign" action. I've [reworded a bit](https://github.com/openpubkey/openpubkey/pull/83/commits/85ca102ebe604104a3bfdaeeb88dcf6e05c794df) and went with standard bullets for easier reading.

## How To Use OpenPubkey

OpenPubkey is driven by its use cases. You can find all available use cases in the [examples folder](./examples/). 

We expect this list to continue growing (and if you have an idea for an additional use case, please file an issue, raise the idea in a community meeting, or send a message in our Slack channel)!

## How To Develop With OpenPubkey

As we work to get this repository ready for `v 1.0`, you can check out the [examples folder](./examples/) for more information about OpenPubkey's different use cases. In the meantime, we would love for the community to contribute more use cases. See [below](#get-involved-with-our-community) for guidance on joining our community.

## Governance and Contributing

### File An Issue

For feature requests, bug reports, technical questions and requests, please open an issue. We ask that you review [existing issues](https://github.com/openpubkey/openpubkey/issues) before filing a new one to ensure your issue has not already been addressed. 

If you have found what you believe to be a security vulnerability, *DO NOT file an issue*. Instead, please follow our [security disclosure policy](./SECURITY.md).

> ### Code Of Conduct

Reviewer 2: ```suggestion
### Code of Conduct
```

Author: Fixed! [85ca102ebe604104a3bfdaeeb88dcf6e05c794df](https://github.com/openpubkey/openpubkey/pull/83/commits/85ca102ebe604104a3bfdaeeb88dcf6e05c794df)
